google/cloud/bigtable/internal/rate_limiter.h

```diff
@@ -0,0 +1,132 @@
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+
+#include "google/cloud/internal/clock.h"
+#include "google/cloud/log.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <cstdint>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A threadsafe interface for rate limiting.
+ *
+ * The caller needs to acquire a "token" to perform the operation under rate
+ * limits. This class limits the number of tokens issued per period of time,
+ * effectively limiting the operation rate.
+ *
+ * The caller may acquire more than one token at a time if it needs to perform a
+ * burst of the operation under rate limits. More tokens become available as
+ * time passes, with some maximum to limit the size of bursts.
+ *
+ * The allocation of resources must be a "prior reservation". That is, the
+ * caller must tell the `RateLimiter` how many tokens it wants to acquire
+ * *before* performing the operation. The `RateLimiter` will tell the caller
+ * when to perform the operation.
+ *
+ * The `RateLimiter` does not sleep. It is the responsibility of the caller to
+ * sleep. For example:
+ *
+ * @code
+ * auto clock = std::make_shared<internal::SteadyClock>();
+ * auto initial_period = std::chrono::milliseconds(100);
+ * RateLimiter limiter(clock, initial_period);
+ * while (MoreThingsToDo()) {
+ *   auto wait = limiter.acquire(1);
+ *   std::this_thread_sleep_for(wait);
+ *   DoOneThing();
+ * }
+ * @endcode
+ *
+ * Rate limiting does not start until after the first call to `acquire()`.
+ * Consider a caller asking for 100 tokens at 1 token/s. We do not want to wait
+ * 100s for this initial request. Instead, it goes through immediately, and the
+ * next request is scheduled for 100s from now.
+ *
+ * @see https://en.wikipedia.org/wiki/Flow_control_(data)#Open-loop_flow_control
+ */
+class RateLimiter {
+ public:
+  // Absolute time does not matter, so use a steady_clock which is guaranteed to
+  // increase with time.
+  using Clock = ::google::cloud::internal::SteadyClock;
+
+  template <typename Rep1, typename Period1>
+  explicit RateLimiter(std::shared_ptr<Clock> clock,
+                       std::chrono::duration<Rep1, Period1> period)
+      : RateLimiter(clock, period, Clock::duration::zero()) {}
+
+  template <typename Rep1, typename Period1, typename Rep2, typename Period2>
+  explicit RateLimiter(std::shared_ptr<Clock> clock,
+                       std::chrono::duration<Rep1, Period1> period,
+                       std::chrono::duration<Rep2, Period2> smoothing_interval)
+      : clock_(std::move(clock)),
+        // Note that std::chrono::abs() is not available until C++17.
+        smoothing_interval_(std::chrono::duration_cast<Clock::duration>(
+            smoothing_interval >= Clock::duration::zero()
+                ? smoothing_interval
+                : -smoothing_interval)),
```

Comment:
Opt:  Maybe a private, static `abs()` helper to eliminate the duplication?

Reply:
That is significantly nicer. Thanks.